### PR TITLE
Typescript to create the json and swift bindings

### PR DESCRIPTION
**Pull request by langleyd**
- json generation typescript
- Adds emoji.ts from matrix-react-sdk
- Adds swift bindings + tests + package

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "compilerOptions": {
+    "target": "es2016",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
+    "module": "commonjs",                                /* Specify what module code is generated. */
+    "esModuleInterop": true,                             /* Emit additional JavaScript to ease support for importing CommonJS modules. This enables 'allowSyntheticDefaultImports' for type compatibility. */
+    "forceConsistentCasingInFileNames": true,            /* Ensure that casing is correct in imports. */
+    "strict": true,                                      /* Enable all strict type-checking options. */                   /* Skip type checking .d.ts files that are included with TypeScript. */
```
Comment (t3chguy) [inline tsconfig.json:7]: `/* Skip type checking .d.ts files that are included with TypeScript. */` this comment seems unrelated to anything here

File: platforms/ios/Emojibase/Resources/emojibase.json [hunk 2]
```diff
@@ -0,0 +1 @@
...1F46D-1F3FD","label":"women holding hands: medium skin tone","order":2678,"unicode":"👭🏽"},{"group":1,"hexcode":"1F46D-1F3FE","label":"women holding hands: medium-dark skin tone","order":2684,"unicode":"👭🏾"},{"group":1,"hexcode":"1F46D-1F3FF","label":"women holding hands: dark skin tone","order":2690,"unicode":"👭🏿"},{"group":1,"hexcode":"1F469-1F3FB-200D-1F91D-200D-1F469-1F3FC","label":"women holding hands: light skin tone, medium-light skin tone","order":2667,"unicode":"👩🏻‍🤝‍👩🏼"},{"group":1,"hexcode":"1F469-1F3FB-200D-1F91D-200D-1F469-1F3FD","label":"women holding hands: light skin tone, medium skin tone","order":2668,"unicode":"👩🏻‍🤝‍👩🏽"},{"group":1,"hexcode":"1F469-1F3FB-200D-1F91D-200D-1F469-1F3FE","label":"women holding hands: light skin tone, medium-dark skin tone","order":2669,"unicode":"👩🏻‍🤝‍👩🏾"},{"group":1,"hexcode":"1F469-1F3FB-200D-1F91D-200D-1F469-1F3FF","label":"women holding hands: light skin tone, dark skin tone","order":2670,"unicode":"👩🏻‍🤝‍👩🏿"},{"group":1,"hexcode":"1F469-1F3FC-200D-1F91D-200D-1F469-1F3FB","label":"women holding hands: medium-light skin tone, light skin tone","order":2671,"unicode":"👩🏼‍🤝‍👩🏻"},{"group":1,"hexcode":"1F469-1F3FC-200D-1F91D-200D-1F469-1F
```
Comment (t3chguy) [inline platforms/ios/Emojibase/Resources/emojibase.json:1]: Keeping this in the repo isn't ideal, is this needed for swift packages to function? Can we not build it and deploy it into a package repo ala npm for swift?

File: package.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
+  "devDependencies": {
+    "ts-node": "^10.9.1",
+    "typescript": "^5.1.6"
+  },
+  "dependencies": {
+    "@types/node": "^20.4.6",
```
Comment (t3chguy) [inline package.json:7]: this is a devDep

File: package.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    "emojibase": "^15.0.0",
+    "emojibase-data": "^15.0.0",
+    "i": "^0.3.7",
```
Comment (t3chguy) [inline package.json:10]: What is this?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    "npm": "^9.8.1"
```
Comment (t3chguy) [inline package.json:11]: We should be using `yarn` and not including a package manager as a dev dependency as it cause issues on dev boxes

**Review by t3chguy — COMMENTED** [review]
We should set up renovate to ensure we keep emojibase up to date too
As well as github workflow automations to run the tests & handle releases

File: platforms/ios/Emojibase/Resources/emojibase.json (hunk 2 above)
Comment (langleyd) [reply platforms/ios/Emojibase/Resources/emojibase.json:1]: SwiftPM deals directly with git repos, so the layout here would be common place. We could write a script to compile it to an xcframework that gets uploaded to the release binary but it's more common to ship the source with your package. Any thoughts @stefanceriu?

File: package.json (hunk 4 above)
Comment (langleyd) [reply package.json:10]: I *don't* know, ha. Will see if I can remove it.

Comment (t3chguy) [reply package.json:10]: I'm guessing it was an accident of `yarn install i <blah>` where `i` is the short alt for `install`

**Comment by langleyd** [comment]
@t3chguy any chance you would be able to add the release/test automation when adding the npm publishing? I see you did a very nice job of [matrix-analytics-events](https://github.com/matrix-org/matrix-analytics-events/blob/main/.github/workflows/release.yaml) :).

On the swift side we've get publishing/versioning for free as long as the github releases are created.

I can tag this manually after this PR and that will allow me integrate the on element-x-ios. I'll get to work on the kotlin bindings then.

**Comment by t3chguy** [comment]
> On the swift side we've get publishing/versioning for free as long as the github releases are created.

Is it about the github release or just the git tag?

> @t3chguy any chance you would be able to add the renovate and release/test automation when adding the npm publishing? I see you did a very nice job of [matrix-analytics-events](https://github.com/matrix-org/matrix-analytics-events/blob/main/.github/workflows/release.yaml?rgh-link-date=2023-08-04T13%3A43%3A04Z) :).

Sure

**Review by t3chguy — CHANGES_REQUESTED** [review]
Think we should have a github workflow linter to prevent the emojibase.json from falling out of sync

**Comment by langleyd** [comment]
> Is it about the github release or just the git tag?
Sorry yea just the tag.

**Comment by langleyd** [comment]
> Think we should have a github workflow linter to prevent the emojibase.json from falling out of sync

Do you mean using renovate to stop it getting out of sync with the upstream emojibase?

Or something that validates the json against a schema?

Or something else?

File: .github/workflows/validate-json.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+name: Validate emojibase.json
+
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+
+jobs:
+  build:
+    runs-on: ubuntu-20.04
```
Comment (t3chguy) [inline .github/workflows/validate-json.yaml:10]: Any reason to not use `ubuntu-latest` here?

File: .github/workflows/validate-json.yaml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    concurrency:
+      # When running on main, use the sha to allow all runs of this workflow to run concurrently.
+      # Otherwise only allow a single run of this workflow on each branch, automatically cancelling older runs.
+      group: ${{ github.ref == 'refs/heads/main' && format('validate-json-main-{0}', github.sha) || format('validate-json-pr-{0}', github.ref) }}
+      cancel-in-progress: true
```
Comment (t3chguy) [inline .github/workflows/validate-json.yaml:16]:
```suggestion
      # When running on main, use the sha to allow all runs of this workflow to run concurrently.
      # Otherwise only allow a single run of this workflow on each branch, automatically cancelling older runs.
      group: ${{ github.workflow }}
      cancel-in-progress: ${{ github.ref_name != 'main' }}
```

This seems simpler yet and has the benefit of not running an unbound number of `main` branch builds in parallel causing runner starvation for the whole org.

File: .github/workflows/validate-json.yaml (hunk 6 above)
Comment (langleyd) [reply .github/workflows/validate-json.yaml]: No, was a copy/paste :)

**Review by t3chguy — APPROVED** [review]

File: platforms/ios/Emojibase/Resources/emojibase.json (hunk 2 above)
Comment (stefanceriu) [reply platforms/ios/Emojibase/Resources/emojibase.json:1]: Another option might be using a package plugin instead but I don't think either of them are worth the effort, writing or maintaining.
